File: project/block_manager/services/nodes/frameworks/pytorch/nodes/output.py
"""PyTorch Output Node Definition"""

from typing import Any, Dict, List, Optional

from ....core.types import ConfigField, Framework, NodeMetadata, TensorShape
from ..base import PyTorchTerminalNode


class OutputNode(PyTorchTerminalNode):
    """Output layer for the neural network"""

    @property
    def metadata(self) -> NodeMetadata:
        return NodeMetadata(
            type="output",
            label="Output",
            category="output",
            color="var(--color-green)",
            icon="Export",
            description="Define model output and predictions",
            framework=Framework.PYTORCH
        )

    @property
    def config_schema(self) -> List[ConfigField]:
        return []  # No configuration needed

    def compute_output_shape(
        self,
        input_shape: Optional[TensorShape],
        config: Dict[str, Any]
    ) -> Optional[TensorShape]:
        # Output shape is same as input
        return input_shape

File: packages/core/src/lib/nodes/definitions/pytorch/output.ts
/**
 * PyTorch Output Node Definition
 */

import { TerminalNodeDefinition } from '../../base'
import { NodeMetadata, BackendFramework } from '../../contracts'
import { TensorShape, BlockConfig, ConfigField } from '../../../types'
import { PortDefinition } from '../../ports'

export class OutputNode extends TerminalNodeDefinition {
  readonly metadata: NodeMetadata = {
    type: 'output',
    label: 'Output',
    category: 'output',
    color: 'var(--color-green)',
    icon: 'Export',
    description: 'Define model output and predictions',
    framework: BackendFramework.PyTorch
  }

  readonly configSchema: ConfigField[] = []

  /**
   * Output node provides predictions that can connect to loss functions
   */
  getOutputPorts(config: BlockConfig): PortDefinition[] {
    return [{
      id: 'predictions-output',
      label: 'Predictions',
      type: 'output',
      semantic: 'predictions',
      required: false,
      description: 'Model predictions/output'
    }]
  }

  computeOutputShape(inputShape: TensorShape | undefined, config: BlockConfig): TensorShape | undefined {
    return inputShape
  }
}
